### ISSUE TITLE
Add pre validate hook throw exception during after validation if there is errors

### ISSUE THREAD

**Issue by julien-maurel** [LearnBoost/mongoose#2567]
Hi all,
I have an issue with pre validate hook.
If I try to save a document without this kind of hook and a field have an error, no worries, save callback is called with err.
But if I add a pre validate hook, instead to have callback called, it throw an exception.

I posted an issue (2484) about a similar error during model.create.

Thanks.

**Comment by vkarpov15** [comment]
Hmm can you provide a standalone example that reproduces this? I don't quite understand.

**Comment by julien-maurel** [comment]
Basically, if I do doc.save(function(err){...}), callback is always called, normal.
But, if I have a pre validate hook (like below) on the model and that during the save, a validator is not valid, instead to go on callback with err that is equal to the error, an exception is thrown.

```
// Schema
schema.add({field:{type:String, required:true})
...
// Save
doc=new MyDoc();
doc.save(function(err){
}); // Callback is called with err because field is missing and required
```

```
// Schema
schema.add({field:{type:String, required:true})
...
// Add pre validate hook
schema.pre('validate',function(next){
  console.log('OK');
  next();
});
// Save
doc=new MyDoc();
doc.save(function(err){
}); // Console.log is done and callback should be called like before, but not, instead, exception is thrown
```

**Comment by julien-maurel** [comment]
I didn't precise I use 3.9 version, I don't know if it is the same thing on 3.8

**Comment by julien-maurel** [comment]
More, from 3.9.5, there is below code on lib/schema.js around line 119. So with taht, there is always a pre validate hook and when I use doc.save with errors, there is always exception instead callback.
So fix this pre validate error will fix save ;)
```
  this.pre('validate', function(next) {
    delete this.errors;
    next();
  });
```

**Comment by julien-maurel** [comment]
I found the issue.

If pre validator is added on a method (save, validate...), orignal method is overwrite by $__registerHooksFromSchema method on lib/document.js.
On this method, at line 1499, you add a reject listener via $__handleReject that emit error event if there is error listener and else, thrown an exception.
If I add a "doc.on('error', function)", event is called and callback too because $__handleReject doesn't no longer thrown an exception.

I don't know if it is a good fix, but I did below changes on lib/document
```
      if (typeof lastArg == 'function') {
        wrapingPromise.onResolve(lastArg);
      }
      if (!(this instanceof Embedded) && !wrapingPromise.hasRejectListeners()) {
        wrapingPromise.onReject(self.$__handleReject.bind(self));
      }
      args.push(function () {
        return wrapingPromise.resolve.apply(wrapingPromise, arguments);
      });
```
to 
```
      if (typeof lastArg == 'function') {
        wrapingPromise.onResolve(lastArg);
      }
      args.push(function () {
        if (!(this instanceof Embedded) && !wrapingPromise.hasRejectListeners()) {
          wrapingPromise.onReject(self.$__handleReject.bind(self));
        }
        return wrapingPromise.resolve.apply(wrapingPromise, arguments);
      });
```

With that, it give possibility to add reject listener on the promise and that fix pre validate error.

**Comment by vkarpov15** [comment]
Thanks for the detailed report, that'll help me a lot in fixing this

**Comment by alancasagrande** [comment]
I'm experiencing the same issue. Quick code to reproduce:

```javascript
var mongoose = require('mongoose');

var schema = mongoose.Schema({
  name: { type: String, required: true }
});

var User = mongoose.model('User', schema);

new User({ name: 'user' }).validate().then(function () {
  console.log('works ok');

  new User().validate().then(null, function (err) {
    console.log('this error callback is not called, an exception is thrown instead');
  });
});
```

**Comment by julien-maurel** [comment]
Hi,
I tried the last release (4.0.0-rc1), and it works now. So I think this issue is fixed in 4.0.0

**Comment by vkarpov15** [comment]
Actually, this doesn't happen at all in 3.8.x, this was introduced in 3.9.1 and went away thanks to some unrelated work in 4.0.0-rc0. Thanks for pointing this out :)
